src/Interpreters/ActionsVisitor.cpp

```diff
@@ -945,12 +963,45 @@ void ActionsMatcher::visit(const ASTFunction & node, const ASTPtr & ast, Data &
         // Don't need to do anything more for window functions here -- the
         // resulting column is added in ExpressionAnalyzer, similar to the
         // aggregate functions.
+        if (data.window_dependancy_state == WindowDependancyState::MAY_DEPEND)
+            data.window_function_in_subtree = true;
         return;
     }
+    else if (node.compute_after_window_functions)
```

Comment:
So, this flags `subtree_contains_window_call`, `compute_after_window_functions`, `window_function_in_subtree`, `window_dependancy_state` are required just to change the order of visiting the nodes, are they? It looks like this PR's trickiest part, but it just does `visit`.

Reply:
Yes, that's the most important part of this PR. Let's consider some general expression with a window function:
```
f(g(arg) OVER window_name)
```
Evaluation of this expression must be split to 3 separate parts:
1. Evaluate all arguments of window functions (`arg` in the example)
2. Evaluate window functions. It's important to notice, that any window function can't have another window function as an argument. (In the example it's `g(arg) OVER window_name`)
3. Evaluate expressions with window functions (in the example we will compute`f` call value). We can do this way because we are sure that `f` is not:
    * Aggregate function (aggregation is executed before window functions, so I can't have a window function as an argument);
    * Window function.

In other words, here we have some kind of _topological sort_.